File: CSharp/Shop/Shop.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;

namespace Shop {
    using Items;

    public class Shop {
        private int revenue = 0;

        private List<ItemBatch> items = new List<ItemBatch>();
        private List<ItemBatch> soldItems = new List<ItemBatch>();

        public List<ItemBatch> Items {
            get { return items; }
        }
        public int Revenue {
            get { return revenue; }
        }
        public void Add(ItemBatch batch) {
            bool found = false;
            for(int i = 0; i < items.Count; ++i) {
                if (items[i].Item == batch.Item) {
                    items[i].Stock += batch.Stock;
                    found = true;
                }
            }
            if (!found) {
                items.Add(batch);
            }
        }
        public int StockAvailable(string itemName) {
            for(int i = 0; i < items.Count; ++i) {
                if (items[i].Item.Name == itemName) {
                    return items[i].Stock;
                }
            }
            return 0;
        }
        public Item Buy(string item, ref int funds) {
            for(int i = 0; i < items.Count; ++i) {
                if (items[i].Item.Name == item) {
                    Item currentItem = items[i].Item;
                    if(currentItem.Value > funds) {
                        throw new NotEnoughFundsException();
                    }
                    SoldItem(currentItem, ref funds);
                    return currentItem;
                }
            }
            throw new ItemNotAvailableException();
        }
        public void SoldItem(Item item, ref int funds) {
            funds -= item.Value;
            revenue += item.Value;

            RemoveItem(item);
            AddSoldItem(item);
        }
        private void RemoveItem(Item item) {
            for(int j = 0; j < items.Count; ++j) {
                if (items[j].Item == item) {
                    items[j].Stock--;
                }
                if (items[j].Stock <= 0) {
                    items.RemoveAt(j);
                }
            }
        }
        private void AddSoldItem(Item item) {
            bool found = false;
            for(int j = 0; j < soldItems.Count; ++j) {
                if (soldItems[j].Item == item) {
                    soldItems[j].Stock++;
                    found = true;
                }
            }
            if (!found) {
                soldItems.Add(new ItemBatch(item, 1));
            }
        }
        public void Return(Item item) {
            for(int i = 0; i < soldItems.Count; ++i) {
                if (soldItems[i].Item == item) {
                    revenue -= item.Value;
                    AddSingleItemToStock(item);
                    soldItems.RemoveAt(i);
                    return;
                }
            }
            throw new CannotReturnException();
        }
        private void AddSingleItemToStock(Item item) {
            bool found = false;
            for(int j = 0; j < items.Count; ++j) {
                if (items[j].Item == item) {
                    items[j].Stock++;
                    found = true;
                }
            }
            if (!found) {
                items.Add(new ItemBatch(item, 1));
            }
        }
    }

    public class NotEnoughFundsException : Exception{
        public NotEnoughFundsException() : base("Not enough Funds!"){
        }
    }
    public class ItemNotAvailableException : Exception{
        public ItemNotAvailableException() : base("Item not available!"){
        }
    }
    public class CannotReturnException : Exception{
        public CannotReturnException() : base("Item was never bought!"){
        }
    }
}

File: Java/TddShop/src/Shop/Shop.java
package Shop;

import Item.Item;
import Item.ItemBatch;

import java.util.ArrayList;
import java.util.List;

import Customer.Wallet;

public class Shop {
    private List<ItemBatch> items;
    private List<Item> soldItems;
    private int revenue = 0;
    public Shop(){
        items = new ArrayList<ItemBatch>();
        soldItems = new ArrayList<Item>();
    }
    public void Add(ItemBatch b){
        for(int i = 0; i < items.size(); ++i){
            if(items.get(i).Item().Name().equalsIgnoreCase(b.Item().Name())){
                items.get(i).Stock += b.Stock;
                return;
            }
        }
        items.add(b);
    }
    public List<ItemBatch> Items(){
        return items;
    }
    public int StockAvailable(String itemname){
        for(int i = 0; i < items.size(); ++i){
            if(itemname.equalsIgnoreCase(items.get(i).Item().Name())){
                return items.get(i).Stock;
            }
        }
        return 0;
    }
    public void DecrementStock(int itemIndex){
        --items.get(itemIndex).Stock;
        if(items.get(itemIndex).Stock <= 0){
            items.remove(itemIndex);
        }
    }
    public Item Buy(String itemName, Wallet wallet) throws NotEnoughFundsException, ItemNotAvailableException {
        boolean found = false;
        for(int i = 0; i < items.size(); ++i){
            if(items.get(i).Item().Name().equalsIgnoreCase(itemName)){
                found = true;
                if(wallet.Funds >= items.get(i).Item().Value()){
                    Item item = items.get(i).Item();
                    wallet.Funds -= items.get(i).Item().Value();
                    revenue += items.get(i).Item().Value();
                    soldItems.add(item);
                    DecrementStock(i);
                    return item;
                }
                else{
                    throw new NotEnoughFundsException();
                }
            }
        }
        throw new ItemNotAvailableException();
    }
    public void Return(Item item) throws CannotReturnException{
        boolean found = false;
        for(int i = 0; i < soldItems.size(); ++i){
            if(soldItems.get(i).Name().equalsIgnoreCase(item.Name())){
                found = true;
                soldItems.remove(i);
            }
        }
        if(!found){
            throw new CannotReturnException();
        }
        found = false;
        for(int i = 0; i < items.size(); ++i){
            if(items.get(i).Item().Name().equalsIgnoreCase(item.Name())){
                items.get(i).Stock++;
                found = true;
            }
        }
        if(!found){
            ItemBatch b = new ItemBatch(item, 1);
            items.add(b);
        }
        revenue -= item.Value();
    }
    public int Revenue(){
        return revenue;
    }
}
